File: go/classic/minSubArrayLen_209.go
package classic

import "math"

func minSubArrayLenWithTwoPointer(target int, nums []int) int {
	sum := 0
	i, j := -1, -1
	minLen := math.MaxInt32
	for i <= j && (i < len(nums) || j < len(nums)) {
		for sum <= target {
			if sum >= target {
				minLen = min(minLen, j-i)
			}
			j++
			if j >= len(nums) {
				break
			}
			sum += nums[j]
		}

		if j >= len(nums) && sum < target {
			break
		}

		for sum >= target {
			if sum >= target {
				minLen = min(minLen, j-i)
			}
			i++
			if i >= len(nums) {
				break
			}
			sum -= nums[i]
		}
	}
	if minLen == math.MaxInt32 {
		return 0
	}
	return minLen
}

// 找到 nums 中的一个子数组其和大于等于 target
func minSubArrayLen(target int, nums []int) int {
	sum := 0
	left := 0
	minLen := math.MaxInt32

	for i, num := range nums {
		sum += num
		for sum >= target {
			minLen = min(minLen, i-left+1)
			sum -= nums[left]
			left++
		}
	}
	if minLen == math.MaxInt32 {
		return 0
	}
	return minLen
}

File: java/src/MinSubArrayLen_209.java
public class MinSubArrayLen_209 {

    public int minSubArrayLen(int target, int[] nums) {
        int left = 0, right = 0;
        int sum = 0;
        int res = Integer.MAX_VALUE;

        while (right < nums.length) {
            sum += nums[right++];
            while (left <= right && sum >= target) {
                res = Math.min(res, right - left);
                sum -= nums[left++];
            }
        }

        return res == Integer.MAX_VALUE ? 0 : res;
    }

    public static void main(String[] args) {
        System.out.println(new MinSubArrayLen_209().minSubArrayLen(7, new int[]{2, 3, 1, 2, 4, 3}));
        System.out.println(new MinSubArrayLen_209().minSubArrayLen(4, new int[]{1, 4, 4}));
        System.out.println(new MinSubArrayLen_209().minSubArrayLen(11, new int[]{1, 2, 3, 4, 5}));
    }
}
